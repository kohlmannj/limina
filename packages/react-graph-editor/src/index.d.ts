export type PointTuple = [number, number];

export type PointShape = 'circle' | 'diamond' | 'square';

export type OriginDirection = 'top' | 'bottom' | 'left' | 'right';

export type ScrollBarAxis = 'x' | 'y';

export type OverflowMode = 'scroll' | 'auto';

export interface IViewBox {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
  width: number;
  height: number;
  originX?: OriginDirection;
  originY?: OriginDirection;
}

export interface IDomainRange {
  domain: PointTuple;
  range: PointTuple;
}

export interface IThemeProps {
  color: string;
  thickness: number;
  shape: PointShape;
}
