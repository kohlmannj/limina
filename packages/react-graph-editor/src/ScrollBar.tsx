import React, { CSSProperties, SFC } from 'react';
import styled, { css, StyledComponent } from 'react-emotion';
import { OverflowMode, ScrollBarOrientation } from './index';

export interface IScrollBarProps {
  className?: string;
  orientation?: ScrollBarOrientation;
  overflow?: OverflowMode;
  progress?: number;
  scale?: number;
  style?: CSSProperties;
}

export interface IScrollBarTheme {
  thumbColor?: string;
  thumbWidth?: number;
  trackColor?: string;
  trackWidth?: number;
}

const getOvershootFactor = (progress = 0, min = 0, max = 1): number => {
  let overshoot;

  if (progress < min) {
    overshoot = 1 - Math.abs(progress);
  } else if (progress >= max) {
    overshoot = 1 - Math.abs(progress - max);
  } else {
    overshoot = 1;
  }

  return overshoot;
};

const positionForOrientation = (props: IStyledScrollBarProps) => css`
  left: ${!props.orientation || props.orientation === 'vertical' ? 'auto' : 0};
  right: ${!props.orientation || props.orientation === 'vertical'
    ? 0
    : `${
        props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
      }px`};
  top: ${!props.orientation || props.orientation === 'vertical' ? 0 : 'auto'};
  bottom: ${!props.orientation || props.orientation === 'vertical'
    ? `${props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`
    : 0};
`;

const Track = styled<IScrollBarProps, 'div'>('div')`
  background: ${props => props.theme.trackColor || `#efefef`};
  display: flex;
  flex-direction: ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'column' : 'row'};
  align-items: center;
  justify-content: ${props =>
    typeof props.progress === 'number' && props.progress >= 0.5 ? 'flex-end' : 'flex-start'};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`};
  ${positionForOrientation};
`;

const ThumbContainer = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: inherit;
  justify-content: inherit;
  align-items: center;
  --overshoot-factor: ${props => getOvershootFactor(props.progress)};
  ${props => (!props.orientation || props.orientation === 'vertical' ? 'width' : 'height')}: 100%;
  /* Thumb Length */
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'height' : 'width'}: ${props =>
    `calc(1 / ${props.scale || 1} * ${getOvershootFactor(props.progress)} * 100%)`};
  transform: ${props =>
    `translate${!props.orientation || props.orientation === 'vertical' ? 'Y' : 'X'}(${(Math.min(
      Math.max(props.progress || 0, 0),
      1
    ) +
      (typeof props.progress === 'number' && props.progress >= 0.5 ? -1 : 0)) *
    ((props.scale || 1) -
      1) /* *
      (typeof props.progress === 'number' && props.progress > 1
        ? (props.progress || 1) / getOvershootFactor(props.progress)
        : 1) */ *
      100}%)`};
`;

export interface IStyledScrollBarProps extends IScrollBarProps {
  theme?: IScrollBarTheme;
}

export const Thumb: StyledComponent<IScrollBarProps, IScrollBarTheme, {}> = styled<
  IScrollBarProps,
  'button'
>('button')`
  background: ${props => props.theme.thumbColor || '#666'};
  border-radius: ${props =>
    `${(typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15) / 2}px`};
  display: flex;
  padding: 0;
  border: 0;
  flex-direction: inherit;
  flex: 0 1 100%;
  margin: ${props =>
    `${!props.orientation || props.orientation === 'vertical' ? '' : '0 '}${((typeof props.theme
      .trackWidth === 'number'
      ? props.theme.trackWidth
      : 15) -
      (typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15)) /
      2}px${!props.orientation || props.orientation === 'vertical' ? ' 0' : ''}`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15}px`};
`;

export const ScrollBar: SFC<IScrollBarProps> = ({
  className,
  orientation,
  overflow,
  progress,
  scale,
  style,
}) => (
  <Track className={className} orientation={orientation} progress={progress} style={style}>
    <ThumbContainer orientation={orientation} progress={progress} scale={scale}>
      {typeof scale === 'number' &&
        (scale > 1 || (scale <= 1 && overflow === 'scroll')) && <Thumb orientation={orientation} />}
    </ThumbContainer>
  </Track>
);

ScrollBar.defaultProps = {
  orientation: 'vertical',
  overflow: 'auto',
  scale: 1,
};

export default ScrollBar;
