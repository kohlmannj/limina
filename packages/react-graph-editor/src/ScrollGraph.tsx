import React, { Component, CSSProperties, SFC } from 'react';
import styled, { css } from 'react-emotion';
import { withProps } from 'recompose';
import Graph, { IGraphProps } from './Graph';

export interface IScrollGraphProps extends IGraphProps {
  scaleX?: number;
  scaleY?: number;
  style?: CSSProperties;
}

const HorizontalRange = styled('input')`
  width: 150px;
  height: 15px;
  margin: 0;
  position: absolute;
  left: 0;
  bottom: 0;
  z-index: 1;
  transition: 0.25s opacity ease-in-out;
  opacity: 0;

  &:hover,
  &:focus {
    opacity: 1;
  }
`;

const VerticalRange = styled(HorizontalRange)`
  transform-origin: 7.5px 50%;
  transform: rotate(90deg);
  top: 0;
  right: -135px;
  bottom: auto;
  left: auto;
`;

const GraphContainer = styled.div`
  width: 100%;
  height: 100%;
  overflow: auto;
  background: rgba(0, 0, 0, 0.25);
`;

const GraphTrack = styled.div`
  box-sizing: border-box;
  padding-right: 15px;
  padding-bottom: 15px;
`;

const ScrollGraph: SFC<IScrollGraphProps> = ({ className, scaleX, scaleY, style, ...rest }) => (
  <GraphContainer className={className} style={style}>
    <GraphTrack
      className={css({
        width: `${scaleX! * 100}%`,
        height: `${scaleY! * 100}%`,
      })}
    >
      <Graph {...rest} />
    </GraphTrack>
    <HorizontalRange type="range" value={scaleX} min={1} max={10} />
    <VerticalRange type="range" value={scaleY} min={1} max={10} />
  </GraphContainer>
);

ScrollGraph.defaultProps = {
  scaleX: 1,
  scaleY: 1,
};

export default ScrollGraph;
