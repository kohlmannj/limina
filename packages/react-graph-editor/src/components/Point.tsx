import React, { CSSProperties, SFC } from 'react';
import styled, { css, cx, StyledComponent } from 'react-emotion';
import { IThemeProps, IViewBox, OriginDirection, PointShape, PointTuple } from '../index';
import ViewBoxContext from './ViewBoxContext';

export interface IPointProps {
  className?: string;
  selected?: boolean;
  theme?: Partial<IThemeProps>;
  x: number;
  y: number;
}

export interface IStyledButtonProps extends IPointProps, IViewBox {}

const getOriginTranslateFactor = (origin: OriginDirection) =>
  origin === 'left' || origin === 'top' ? -0.5 : 0.5;

export type DynamicStyle = (props: IPointProps) => string;

export const shapes: Record<PointShape, DynamicStyle> = {
  circle: () => css`
    border-radius: 50%;
  `,
  diamond: () => '',
  square: () => '',
};

const selectedClassName = (props: IPointProps) => css`
  background: #fff;
  border-color: ${props.theme && props.theme.color ? props.theme.color : '#000'};
  box-shadow: 0 0 5px 2px rgba(59, 153, 252, 0.75);
`;

const translate = (props: IStyledButtonProps) =>
  props.originX && props.originY
    ? `translate(${getOriginTranslateFactor(props.originX) * 100}%, ${getOriginTranslateFactor(
        props.originY
      ) * 100}%)`
    : '';

const position = (props: IStyledButtonProps) => css`
  position: absolute;
  ${props.originX && `${props.originX}: ${(props.x - props.minX) / props.width * 100}%;`}
  ${props.originY && `${props.originY}: ${(props.y - props.minY) / props.height * 100}%;`}
  transform: ${translate(props)} ${
  props.theme && props.theme.shape === 'diamond' ? 'rotate(45deg)' : ''
};
`;

const Button: SFC<IStyledButtonProps> = ({ className, selected, x, y, ...rest }) => (
  <button
    className={className}
    aria-label={`Point (${x}, ${y})`}
    title={`(${x}, ${y})`}
    {...rest}
  />
);

const StyledButton: StyledComponent<IStyledButtonProps, IThemeProps, {}> = styled(Button)`
  width: ${props => `${props.theme.thickness || 8}px`};
  height: ${props => `${props.theme.thickness || 8}px`};
  box-sizing: border-box;
  padding: 0;
  border: 1px solid transparent;
  background: ${props => props.theme.color || '#000'};
  box-shadow: 0 0 1px 1px rgba(0, 0, 0, 0.15);
  cursor: move;
  font-size: 0;
  line-height: 0;
  ${props => (!props.theme || props.theme.shape === 'circle') && `border-radius: 50%`};
  ${props => props.selected && selectedClassName};
  ${position};
  outline: 0;
  text-indent: 100%;
`;

export const Point: SFC<IPointProps> = props => {
  return (
    <ViewBoxContext.Consumer>
      {(viewBox: IViewBox) => <StyledButton {...props} {...viewBox} />}
    </ViewBoxContext.Consumer>
  );
};

Point.defaultProps = {
  selected: false,
};

export default Point;
