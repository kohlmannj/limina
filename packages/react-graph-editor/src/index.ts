export * from './components/ScrollBar';
export * from './components/ScrollView';
export * from './components/Graph';
export * from './components/Point';
export * from './components/PolyLineSVG';
export * from './components/ScrollBarCorner';
export * from './components/ScrollGraph';
export * from './components/ViewBoxContext';
export * from './scrollBarTheme';
export * from './utils';

export type PointTuple = [number, number];
export type PointShape = 'circle' | 'diamond' | 'square';
export type OriginDirection = 'top' | 'bottom' | 'left' | 'right';
export type ScrollBarAxis = 'x' | 'y';
export type OverflowMode = 'scroll' | 'auto';

export interface ViewBox {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
  width: number;
  height: number;
  originX?: OriginDirection;
  originY?: OriginDirection;
}

export interface DomainRange {
  domain: PointTuple;
  range: PointTuple;
}

export interface ThemeProps {
  color: string;
  thickness: number;
  shape: PointShape;
}
