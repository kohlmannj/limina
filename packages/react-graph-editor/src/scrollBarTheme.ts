export interface ScrollBarTheme {
  thumbClassName?: string;
  thumbColor?: string;
  thumbWidth?: number;
  trackClassName?: string;
  trackColor?: string;
  trackWidth?: number;
}

export const defaults = {
  thumbColor: '#ccc',
  thumbWidth: 12,
  trackColor: '#333',
  trackWidth: 16,
};
