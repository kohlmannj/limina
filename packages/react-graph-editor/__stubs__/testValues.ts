import { PointTuple } from '../src';
import { IPolyLineSVG } from '../src/PolyLineSVG';

export const testPoints: PointTuple[] = [[320, 320], [480, 0], [1280, 1160]];

export const testLines: IPolyLineSVG[] = [
  { label: 'fontSize', points: [[320, 12], [540, 16], [1280, 24]] },
  { label: 'width', points: [[320, 320], [540, 480], [1280, 1160]] },
  { label: 'height', points: [[320, 320], [540, 405], [1280, 652.5]] },
];
