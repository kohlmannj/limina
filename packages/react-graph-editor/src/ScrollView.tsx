import React, { Component, CSSProperties } from 'react';
import styled, { css } from 'react-emotion';
import { OverflowMode } from './index';
import ScrollBar from './ScrollBar';

export interface IScrollViewProps {
  className?: string;
  dangerouslySetInnerHTML?: {
    __html: string;
  };
  overflow?: OverflowMode;
  overflowX?: OverflowMode;
  overflowY?: OverflowMode;
  style?: CSSProperties;
}

export interface IScrollViewState {
  progressX: number;
  progressY: number;
  scaleX: number;
  scaleY: number;
}

const ScrollBarCorner = styled('div')`
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: 9999;
  width: ${props => props.theme.trackWidth || 15}px;
  height: ${props => props.theme.trackWidth || 15}px;
  background: ${props => props.theme.trackColor || '#efefef'};
`;

const ScrollViewContainer = styled('div')`
  position: relative;
  overflow: hidden;
`;

const ScrollViewOverflowContainer = styled('div')`
  width: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
    }px)`};
  height: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
    }px)`};
  overflow: auto;

  ::-webkit-scrollbar {
    display: none;
  }
`;

export default class ScrollView extends Component<IScrollViewProps, IScrollViewState> {
  public state: IScrollViewState = {
    progressX: 0,
    progressY: 0,
    scaleX: 1,
    scaleY: 1,
  };

  private root?: HTMLDivElement | null;

  private content?: HTMLDivElement | null;

  public componentDidMount() {
    this.startListening();
  }

  public componentWillUnmount() {
    this.stopListening();
  }

  public render() {
    const {
      children,
      className,
      dangerouslySetInnerHTML,
      overflow,
      overflowX,
      overflowY,
      style,
      ...rest
    } = this.props;
    const { progressX, progressY, scaleX, scaleY } = this.state;

    return (
      <ScrollViewContainer
        innerRef={e => {
          this.root = e;
        }}
        className={className}
        style={style}
        {...rest}
      >
        <ScrollViewOverflowContainer
          innerRef={e => {
            this.content = e;
          }}
          dangerouslySetInnerHTML={dangerouslySetInnerHTML}
        >
          {children}
        </ScrollViewOverflowContainer>
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="horizontal"
          overflow={overflowX || overflow}
          progress={progressX}
          scale={scaleX}
        />
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="vertical"
          overflow={overflowY || overflow}
          progress={progressY}
          scale={scaleY}
        />
        <ScrollBarCorner />
      </ScrollViewContainer>
    );
  }

  private onScroll = (): void => {
    if (!this.content) {
      return;
    }

    const { width, height } = this.content.getBoundingClientRect();
    const { scrollTop, scrollHeight, scrollLeft, scrollWidth } = this.content;

    this.setState({
      progressX: scrollLeft / (scrollWidth - width),
      progressY: scrollTop / (scrollHeight - height),
      scaleX: Math.max(scrollWidth / Math.round(width), 1),
      scaleY: Math.max(scrollHeight / Math.round(height), 1),
    });
  };

  private startListening = () => {
    if (this.content) {
      this.content.addEventListener('scroll', this.onScroll);
      this.content.addEventListener('resize', this.onScroll);
      this.onScroll();
    }
  };

  private stopListening = () => {
    if (this.content) {
      this.content.removeEventListener('scroll', this.onScroll);
      this.content.removeEventListener('resize', this.onScroll);
    }
  };
}
