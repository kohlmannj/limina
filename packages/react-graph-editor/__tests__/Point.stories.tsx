import { storiesOf } from '@storybook/react'; // tslint:disable-line no-implicit-dependencies
import { ThemeProvider } from 'emotion-theming';
import React from 'react';
import { testPoints } from '../__stubs__/testValues';
import { PointTuple } from '../src/index';
import Point from '../src/Point';

storiesOf('react-graph-editor/Point', module)
  .add('with default appearance', () => <Point x={0} y={0} />)
  .add('with selected appearance', () => <Point selected x={0} y={0} />)
  .add('with diamond appearance', () => (
    <ThemeProvider theme={{ shape: 'diamond' }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with diamond, selected appearance', () => (
    <ThemeProvider theme={{ shape: 'diamond' }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with square appearance', () => (
    <ThemeProvider theme={{ shape: 'square' }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with square, selected appearance', () => (
    <ThemeProvider theme={{ shape: 'square' }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with theme', () => (
    <ThemeProvider theme={{ color: 'red', thickness: 16 }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with theme and selected appearance', () => (
    <ThemeProvider theme={{ color: 'red', thickness: 16 }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ));
