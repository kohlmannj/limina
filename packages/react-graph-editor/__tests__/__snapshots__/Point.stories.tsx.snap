// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-graph-editor/Point with default appearance 1`] = `<[object Object] />`;

exports[`react-graph-editor/Point with diamond appearance 1`] = `
<Point
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with diamond, selected appearance 1`] = `
<Point
  selected={true}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with selected appearance 1`] = `<[object Object] />`;

exports[`react-graph-editor/Point with square appearance 1`] = `
<Point
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with square, selected appearance 1`] = `
<Point
  selected={true}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with theme 1`] = `
<Point
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with theme and selected appearance 1`] = `
<Point
  selected={true}
  x={0}
  y={0}
/>
`;
